### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by Parry-Parry**
Significant changes to the package format:
* Reader now accepts prompt + backend
* Backend is any model e.g openai or vllm and its sole purpose is to generate text
* ContextAggregation has been moved outside of prompt to prevent excessive nesting
* LLMAsAJudge implemented though I think moving to a provider will be preferable

### CODE REVIEW

File: pyterrier_rag/_util.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
+from typing import Union, Iterable, Tuple, List, Optional, Any
+import pandas as pd
+import itertools
+
+
+def push_queries(
```
Comment (seanmacavaney) [inline pyterrier_rag/_util.py:6]: how is this different than pt.model.push_queries?

File: pyterrier_rag/backend/_hf.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                "max_new_tokens": self.max_new_tokens,
+                "temperature": 1.0,
+                "do_sample": False,
+                "num_beams": 1,
+            }
+        self._generation_args = generation_args
+        self.model = self._model
+
+    @torch.no_grad()
+    def generate(self, inps: Iterable[str], **kwargs) -> List[str]:
```
Comment (seanmacavaney) [inline pyterrier_rag/backend/_hf.py:65]: check the types here! Shouldn't the output be List[BackendOutput]?

File: pyterrier_rag/_util.py (hunk 1 above)
Comment (Parry-Parry) [reply pyterrier_rag/_util.py:6]: Allows for base column to be changed e.g context_0 / thoughts_0 / qanswer_0

File: pyterrier_rag/backend/_hf.py (hunk 2 above)
Comment (Parry-Parry) [reply pyterrier_rag/backend/_hf.py:65]: Cheers will fix

File: pyterrier_rag/backend/_util.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        from pyterrier_rag.backend._vllm import VLLMBackend
+
+        return VLLMBackend(model_name, **Backend_kwargs)
+    elif backend_type == Backends.OPENAI.value:
+        from pyterrier_rag.backend._openai import OpenAIBackend
+
+        return OpenAIBackend(model_name, **Backend_kwargs)
+    else:
+        raise ValueError(f"Unknown Backend type: {backend_type}")
+
```
Comment (seanmacavaney) [inline pyterrier_rag/backend/_util.py:29]: it would be nice if there was something like a "coerce_backend" method? So if you already had e.g., a VLLM object, it could map it to the associated Backend type

File: pyterrier_rag/backend/_vllm.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+        self._model_name_or_path = model_name_or_path
+        self.model = LLM(model=model_name_or_path, **model_args)
+
+        if generation_args is None:
+            generation_args = {
+                "max_tokens": self.max_new_tokens,
+                "temperature": 1.0,
+            }
+        generation_args["logprobs"] = 20
```
Comment (seanmacavaney) [inline pyterrier_rag/backend/_vllm.py:61]: Is there overhead in always requesting logit outputs? Certainly in terms of memory, yes, right?

Maybe generate should take an optional arg to request them?

File: pyterrier_rag/backend/_hf.py (hunk 2 above)
Comment (seanmacavaney) [inline pyterrier_rag/backend/_hf.py:65]: swallowed kwargs?

File: pyterrier_rag/backend/_hf.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            sliced_outputs = []
+            for output, prompt_length in zip(outputs, prompt_lengths):
+                sliced_outputs.append(output[prompt_length:])
+            outputs = sliced_outputs
+
+        # Decode outputs
+        texts = self.tokenizer.batch_decode(outputs, skip_special_tokens=True)
+
+        return [
+            BackendOutput(text=text, logits=outputs, prompt_length=length)
```
[the remark this answers is not in the document]
Comment (Parry-Parry) [reply pyterrier_rag/backend/_hf.py:98]: Yeah it is not outputting a dict, I can be more explicit though and make it return dict, may be more readable

File: pyterrier_rag/backend/_util.py (hunk 3 above)
Comment (Parry-Parry) [reply pyterrier_rag/backend/_util.py:29]: Ah so a kind of .from_model esque setup

File: pyterrier_rag/backend/_openai.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                if "The response was filtered" in str(e):
+                    print("The response was filtered")
+                    return "ERROR::The response was filtered"
+                time.sleep(0.1)
+                trials -= 1
+        if return_text:
+            completion = completion["choices"][0]["message"]["content"]
+        return completion
+
+    def generate(self, prompt: List[dict], **kwargs) -> List[str]:
```
Comment (seanmacavaney) [inline pyterrier_rag/backend/_openai.py:90]: how does the caller know what kwargs to pass here? It depends on the implementation, no?

File: pyterrier_rag/backend/_vllm.py (hunk 4 above)
Comment (Parry-Parry) [reply pyterrier_rag/backend/_vllm.py:61]: I believe so yes, I think this was for equal output styles but I can make it a kwarg

File: pyterrier_rag/backend/_hf.py (hunk 2 above)
Comment (Parry-Parry) [reply pyterrier_rag/backend/_hf.py:65]: kwargs are not swallowed it was to allow more dynamic generation options e.g o1 / r1 have query-conditioned stopping criteria

File: pyterrier_rag/backend/_openai.py (hunk 6 above)
Comment (Parry-Parry) [reply pyterrier_rag/backend/_openai.py:90]: The notion is simply that if you want to pass additional generation kwargs they go here, I will make this explicit and add documentation

File: pyterrier_rag/_optional.py [hunk 7]
```diff
@@ -1,27 +1,43 @@
...
+        import transformers
+
         return True
     except ImportError:
         return False
 
+
 def is_vllm_availible():
     try:
-        import vllm # noqa: F401
```
[the remark this answers is not in the document]
Comment (Parry-Parry) [reply pyterrier_rag/_optional.py:10]: I currently switched it to importlib get spec which appears to be best practise according to a few different linters / PEP

[the remark this answers is not in the document]
Comment (seanmacavaney) [reply pyterrier_rag/_optional.py:10]: okay, can that be a separate PR? It doesn't seem related to this PR.

File: pyterrier_rag/_util.py (hunk 1 above)
Comment (seanmacavaney) [reply pyterrier_rag/_util.py:6]: ah, gotcha! Maybe a different function name is better to reflect this?

Comment (cmacdonald) [reply pyterrier_rag/_util.py:6]: so this could be upstreamed to pt at some point.

File: pyterrier_rag/backend/_hf.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (seanmacavaney) [reply pyterrier_rag/backend/_hf.py:98]: I mean it looks like `outputs` contains the output IDs, not the logits. (Or maybe it's a data structure that contains both?)

File: pyterrier_rag/backend/_hf.py (hunk 2 above)
Comment (seanmacavaney) [reply pyterrier_rag/backend/_hf.py:65]: ah, my bad, I missed that below

**Comment by cmacdonald** [comment]
can we revert pyterrier_rag/_datasets.py in this PR - it just seems like whitespace/formatting/lint changes

**Comment by Parry-Parry** [comment]
> can we revert the following files in this PR - it just seems like whitespace/formatting/lint changes:
> 
>     * pyterrier_rag/_measures.py
> 
>     * pyterrier_rag/_datasets.py
>       (I'll edit to add more files that seem to be revertable)

Yes sorry @cmacdonald, @seanmacavaney mentioned a few others let me revert.

File: pyterrier_rag/_optional.py [hunk 8]
```diff
@@ -1,27 +1,46 @@
...
+
         return True
-    except ImportError:
+    except ValueError:
         return False
 
+
+def is_transformers_availible():
+    try:
+        find_spec("transformers")
```
Comment (seanmacavaney) [inline pyterrier_rag/_optional.py:15]: I don't think this works as expected. It looks like `find_spec` returns `None` if not found, rather than raising a `ValueError`. https://docs.python.org/3/library/importlib.html#importlib.abc.MetaPathFinder.find_spec

**Comment by cmacdonald** [comment]
```python
from pyterrier_rag.backend import VLLMBackend
from pyterrier_rag.prompt import ContextAggregationTransformer
from pyterrier_rag.readers import Reader

model_name_or_path='mistralai/Mistral-7B-v0.1'
mistral = VLLMBackend(model_name_or_path)

mistral_reader = Reader(mistral)
context_aggregation = ContextAggregationTransformer()

bm25_mistral = bm25_ret %3 >> context_aggregation >> mistral_reader
```

This is the key example from the Mistral notebook. I guess my main point is that I dont know easily what form context aggregation takes place.

**Comment by seanmacavaney** [comment]
> I guess my main point is that I dont know easily what form context aggregation takes place.

I agree. Maybe something like `pyterrier_rag.prompt.Contatenator()` would be clearer?

File: requirements.txt [hunk 9]
```diff
@@ -2,7 +2,10 @@ python-terrier>=0.13.0
 pyterrier_alpha>=0.12.0
 ir_measures>=0.3.4
 regex
-transformers
+transformers==4.44.2
```
Comment (cmacdonald) [inline requirements.txt:5]: do we need this transformers pegging?

**Comment by cmacdonald** [comment]
> As a heads up, it looks like the ghas were previously failing because the dependency caches were somehow corrupt. When I deleted them, that stage of the ghas succeeded again.
> 

good spot
